python debugging module - closes #671 (#673)
bob-crutchley/courseware: topics/python/modules/debugging/README.md

# Debugging

<!--TOC_START-->
## Contents
- [Overview](#overview)
- [What is debugging?](#what-is-debugging)
- [Static analysis](#static-analysis)
- [Dynamic analysis](#dynamic-analysis)
- [Debugging tools](#debugging-tools)
- [Tutorial](#tutorial)
- [Exercises](#exercises)

<!--TOC_END-->
## Overview

This module will be an introduction to debugging with Python.

## What is debugging?

`Debugging` is the process of removing errors from the computer, whether the error is `hardware` or `software` related.
Once this process has been `repeated`, the code is said to be "debugged".

Testing is the process of identifying an error under test conditions. Debugging is the process of removing the error at the root. Testing and debugging should be used in a cyclic nature.

We typically work with both static analysis and dynamic analysis.

## Static analysis

Static analysis is a method of debugging done by examining the code without executing it.
To do this, you need to be able to understand the structure of the code, normally best to do for your own code.
This kind of analysis helps to ensure the code adheres to industry standards.
This can be done manually or using tools.

> Static analysis can reveal errors that would not have revealed themselves until a diaster occurs weeks, months or years later.

Reviewer: ```suggestion
Static analysis can reveal errors that may not have revealed themselves until a disaster occurs weeks, months or years later.
```

## Dynamic analysis

Dynamic analysis is the testing and evaluation of a program by executing data in real time.

The objective of this kind of analysis is to find errors whilst the program is running, rather than examining code offline.
This can also be done manually or using tools.

> Whilst analyzing dynamically you can debug a program in all scenarios for which it was designed.

Author: ```suggestion
Whilst analysing dynamically you can debug a program in all scenarios for which it was designed.
```

It eliminates the need to artificially create situations which can produce errors.
The main problem with this kind of debugging is that it finds bugs but does not prevent them.

## Debugging tools

There are many tools available for debugging, but there is also one built into Python.
This debugging tool is called `pdb`, **P**ython **D**e**B**ugging.
This tool allows you to go through code line by line to identify errors.

`pdb.set_trace()` allows you to set a point you wish to start debugging from.

> 

Reviewer: Think you need a high level description of what pdb commands do (p,n,q and how you run pdb)

> 

Author: ```suggestion
When you run the code with this line inserted into it, the console will come up with the first line of the code. You can navigate through your code using the following:
- `n` - Next line
- `q` - Quit
- `p` - Reference variables 
```

## Tutorial

Create a new python file and enter the following.

```python
import pdb

pdb.set_trace()

a = "aaa"
b = "bbb"
c = "ccc"

final = a + b + c
print(final)
```

Now when you run the program it will show the line, `a = "aaa"`. If you try entering `p a`, you will get an error as you have to pass the line for the variable to be defined so type in `n` and press enter. Then try `p a` again, you should get a return of 'aaa'.

Now if you type `q` and press enter you will quit the program. 

Now run the program again go the final line by typing `n` and enter repeatedly. Now type `p final` and you should get `aaabbbccc` returned.

## Exercises

> Debug the following programs. For the first 2 perform static analysis. For the 3rd and 4th of the exercises use dynamic analysis either manually or through `pdb`.

Reviewer: ```suggestion
Debug the following programs. For the first 2 perform static analysis. For the 3rd and 4th exercises use dynamic analysis either manually or through `pdb`.
```
